### PR TITLE
gateway-api: avoid deep copy when converting http routes

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Shallow-copy `spec.Rules` instead of deep-copying the whole object
Trying to make further optimization based on https://github.com/istio/istio/pull/37139#discussion_r799075531

### CODE REVIEW

**Comment by dddddai** [comment]
Benchmark results on my laptop:
```
name                        old time/op    new time/op    delta
BuildHTTPVirtualServices-4     104µs ± 1%     101µs ± 3%  -2.75%  (p=0.003 n=6+11)

name                        old alloc/op   new alloc/op   delta
BuildHTTPVirtualServices-4    28.1kB ± 0%    26.6kB ± 0%  -5.41%  (p=0.000 n=7+12)

name                        old allocs/op  new allocs/op  delta
BuildHTTPVirtualServices-4       485 ± 0%       455 ± 0%  -6.19%  (p=0.000 n=7+12)
```

The delta can be greater if HTTPRoute spec is larger
@howardjohn Could you please take a look?

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 1]
```diff
@@ -299,12 +287,28 @@ func buildHTTPVirtualServices(obj config.Config, gateways map[parentKey]map[k8s.
...
 		}
 		vs.Route = route
 
 		httproutes = append(httproutes, vs)
+		return nil
+	}
+
+	for _, r := range route.Rules {
+		if len(r.Matches) > 1 {
+			continue
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:300]: nvm, I get it now. Maybe it would be simpler if we remove the splitRules list above and then do it all in one place:

```
	for _, r := range route.Rules {
		if len(r.Matches) > 1 {
			// split the rule to make sure each rule has up to one match
			matches := r.Matches
			for _, m := range matches {
				r.Matches = []k8s.HTTPRouteMatch{m}
				if err := convertHTTPRoute(r); err != nil {
					reportError(err)
					return
				}
			}
		} else {
			if err := convertHTTPRoute(r); err != nil {
				reportError(err)
				return
			}
		}
	}
```

Right now the code is spread out and seems like it could easily break in refactor, etc, and a bit hard to follow. This is also (trivially) more efficient to not store in a list

**Review by howardjohn — APPROVED** [review]
LGTM! One nit but can followup
